File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    move = {"E": [0, 1], "W": [0, -1], "S": [1, 0], "N": [-1, 0] }
    max_h = len(park)
    max_w = len(park[0])
    s = []
    
    for h in range(max_h):
        flag = False
        for w in range(max_w):
            if park[h][w] == "S":
                s = [h, w]
                flag = True
                break
        if flag:
            break
    
    for route in routes:
        route = route.split(" ")
        n = int(route[1])
        dh, dw = move[route[0]]
        flag = False
        
        next_h = s[0] + dh * n
        next_w = s[1] + dw * n
        
        if 0 <= next_h < max_h and 0 <= next_w < max_w:
            if s[0] == next_h:
                start = s[1] if s[1] < next_w else next_w
                end = next_w + 1 if s[1] < next_w else s[1] + 1
                for w in range(start, end):
                    if park[s[0]][w] == "X":
                        flag = True
                        break
                        
            if s[1] == next_w:
                start = s[0] if s[0] < next_h else next_h
                end = next_h + 1 if s[0] < next_h else s[0] + 1
                for h in range(start, end):
                    if park[h][s[1]] == "X":
                        flag = True
                        break
            
            if not flag:
                s = [next_h, next_w]
            
    return s

File: 프로그래머스/1/172928. 공원 산책/공원 산책.js
function solution(park, routes) {
    var answer = [];
    const move = {E: [0, 1], W: [0, -1], S: [1, 0], N: [-1, 0] }
    const w = park.length
    const h = park[0].length
    let position = [];
    for (let i = 0; i < w; i++) {
        for (let j = 0; j < h; j++) {
            if (park[i][j] === "S") {
                position = [i, j]
                break
            }
        }
    }
    
    routes.forEach((it) => {
        const direction = it.split(" ")
        const number = Number(direction[1])
        const [dx, dy] = move[direction[0]]
        let flag = true
        for (let i = 1; i <= number; i++) {
            const nextX = position[0] + dx * i
            const nextY = position[1] + dy * i
            if (nextX < 0 || nextX >= w || nextY < 0 || nextY >= h || park[nextX][nextY] === "X") {
                flag = false
                break
            }     
        }
        if (flag) position = [position[0] + dx * number, position[1] + dy * number]
    })

    return position;
}